### ISSUE TITLE
Debug error > "No call stack available (Internal error: TypeError: Cannot read property '0' of undefined)."

### ISSUE THREAD

**Issue by bgbahoue** [Microsoft/vscode#7683]
- VSCode Version: Version 1.2.1 (1.2.1)
- OS Version: OS X El Capitan Version 10.11.4

Steps to Reproduce:
1. Add a breakpoint
2. Launch debug; it sometimes break 'somewhere' yet I don't have the control and the only message I receive is "No call stack available (Internal error: TypeError: Cannot read property '0' of undefined)."
   See screenshot attached.

Sometimes it works, some other it does, on the same code
<img width="1701" alt="capture d ecran 2016-06-15 a 11 34 57" src="https://cloud.githubusercontent.com/assets/13782433/16064461/3d219950-32ed-11e6-84ff-c0dbd34acd12.png">

**Comment by alechirsch** [comment]
I have having the same exact issue.

- OS Version: OS X El Capitan Version 10.11.5
- VSCode Version: Version 1.2.0 (1.2.0)
![screen shot 2016-06-15 at 1 55 10 pm](https://cloud.githubusercontent.com/assets/6865143/16091311/172d4c1c-3301-11e6-83df-0b765cd1e576.png)

**Comment by maorhayoun** [comment]
+1

**Comment by a624535537** [comment]
+1

**Comment by icezohu** [comment]
+1

**Comment by weinand** [comment]
What version of node are you using?

**Comment by icezohu** [comment]
It is 5.10.0, while vscode version is 1.2.1.

**Comment by weinand** [comment]
Since I'm not sure where this comes from (and I cannot reproduce it myself), I've added additional information to the error message and added some additional checks in the injected code. This will be in the June release (and tomorrow's Insider).

**Comment by bgbahoue** [comment]
I’m at v6.2.0

 

De : Andre Weinand <notifications@github.com>
Répondre à : Microsoft/vscode <reply@reply.github.com>
Date : jeudi 30 juin 2016 21:19
À : Microsoft/vscode <vscode@noreply.github.com>
Cc : Boris Gbahoué <boris@athieme-conseil.fr>, Author <author@noreply.github.com>
Objet : Re: [Microsoft/vscode] Debug error > "No call stack available (Internal error: TypeError: Cannot read property '0' of undefined)." (#7683)

 

Since I'm not sure where this comes from (and I cannot reproduce it myself), I've added additional information to the error message and added some additional checks in the injected code. This will be in the June release (and tomorrow's Insider).

**Comment by sarvaje** [comment]
I'm using VSCode v. 1.4.0-insider (updated right now) on Windows 10  and I have the same problem (tested with node 6.2.2 and node 6.3).

Error message: 'No call stack available ({_command}: Internal error: TypeError: Cannot read property '0' of undefined).

The error is very common, and is almost impossible debug anything :(

**Comment by weinand** [comment]
@sarvaje does the problem occur in VS Code 1.3 too?

**Comment by sarvaje** [comment]
yes, the problem is happening for a long time.
Maybe is not related, but I notice that the problem ocurror more ofen when I try to debug promises (using bluebird for example)

**Comment by Stwissel** [comment]
+1 Have that problem for the first time debugging a command line node app with Q based promises. OS/X current, Node 6.2.2

**Comment by weinand** [comment]
Since I could not reproduce this problem, I've instrumented node-debug so that we can get additional information when the problem occurs.

In the July release (and next Monday's insider release) VS Code should show an almost complete call stack even is this problem occurs.  

![2016-07-16_23-15-11](https://cloud.githubusercontent.com/assets/1898161/16897300/3e3c9bfe-4bad-11e6-85a9-19cc8911b262.png)

Only stackframes that run into the problem will show the error.

If you are affected by this problem please add the following property to your launch configuration: 
`"trace": "7683"`

If the problem occurs again, a stacktrace will appear on the debug console.
Please add this information here so that I can investigate.

**Comment by Stwissel** [comment]
@weinand : depends what is considered "small" :-)
I have a small program that extracts the "url" value from har files (the JSON you get when you record a browser session. I'm sure you can find some of those. The project is a single js file.

package.json

      {
        "name": "harparser",
        "version": "0.0.1",
        "description": "Parse har files to get specific information out",
        "main": "app.js",
        "scripts": {
          "test": "echo \"Error: no test specified\" && exit 1"
        },
        "keywords": [
          "har",
          "browser",
          "api"
        ],
        "author": "Stephan H. Wissel",
        "dependencies": {
          "lodash": "^4.13.1",
          "q": "^1.4.1",
          "querystring": "^0.2.0",
          "stream-json": "^0.4.2",
          "url": "^0.11.0"
        }
      }

app.js

            'use strict';

            var sourceDir = '../rawHAR/';
            var Combo = require('stream-json/Combo');
            var Emitter = require('stream-json/Emitter');
            var _ = require('lodash');
            var Q = require('q');
            var url = require('url');
            var querystring = require('querystring');

            var options = {
                packKeys: true,
                packStrings: true,
                packNumbers: true
            };

            var fs = require('fs');
            var objectCounter = 0;

            var getCount = 0;
            var postCount = 0;
            var urlCollection = [];
            var parameterCollection = [];
            var doneCounter = 0;

            console.log('Start App2');

            var processingArray = [];

            fs.readdir(sourceDir, function (err, data) {
                data.forEach(function (element, counter, reference) {
                    var last = (counter === reference.length - 1);
                    console.log('Processing ' + sourceDir + element + (last ? ' (the last)' : ''));
                    if (element.indexOf('.har') > 0) {
                        // Add one to the list to be processes
                        doneCounter++;
                        processingArray.push(processOneFile(sourceDir + element));
                    }
                });

                console.log('Files are away, results coming in...');

                Q.allSettled(processingArray).then(function (results) {
                    var alltheURLs = _.uniq(urlCollection);
                    var allTheParams = _.uniq(parameterCollection);
                    console.log('Found ', objectCounter, ' objects.');
                    console.log('API URLs with GET:' + getCount);
                    console.log('API URLs with POST:' + postCount);
                    console.log('Unique URLs: ' + alltheURLs.length);
                    console.log('Unique params: ' + allTheParams.length);
                    console.log(JSON.stringify(alltheURLs));
                    console.log(JSON.stringify(allTheParams));
                    // TODO write URLs into file
                });
            });

            //processOneFile(sourceFile);

            function processOneFile(fName) {
                var deferred = Q.defer();
                var combo = new Combo(options);
                var emitter = new Emitter();
                var method = 'unknown';
                var attention = false;
                var methodComing = false;

                var pipeline = fs.createReadStream(fName).pipe(combo).pipe(emitter);

                combo.on('data', function (chunk) {
                    ++objectCounter;
                });
                pipeline.on('keyValue', function (key) {
                    methodComing = (key === 'method');

                    attention = (key === 'url');
                });
                pipeline.on('stringValue', function (stringValue) {
                    if (methodComing) {
                        method = stringValue;
                    }
                    if (attention) {
                        if (method == 'GET') {
                            getCount++;
                        }
                        if (method == 'POST') {
                            postCount++;
                        }
                        // Capture the URL
                        updateURLCollection(stringValue);

                    }
                    attention = false;

                });
                pipeline.on('finish', function () {
                    doneCounter--;
                    console.log('Finished ' + fName + ' (' + doneCounter + ')');
                    deferred.resolve(fName);
                });

                return deferred.promise;
            }

            // Rip the queries apart
            function updateURLCollection(rawURL) {
                var sourceURL = url.parse(rawURL);
                // Capture the base URL
                urlCollection.push(sourceURL.host + sourceURL.pathname);
                // now process the parameters
                var rawQuery = sourceURL.query;
                if (rawQuery) {
                    var query = querystring.parse(rawQuery);
                    for (var param in query) {
                        parameterCollection.push(param);
                    }
                }
            }

That one shows the error. To test it, you need some [har files](https://www.maxcdn.com/one/visual-glossary/har-file/) in the <code>../rawHAR/</code> directory.

**Comment by weinand** [comment]
@Stwissel thanks a lot for the code.

Here is what I did to try to reproduce the issue:
- with your files I've created project with VS Code 1.3.1, node 6.2.2 on OS X 10.11.5
- ran `npm install`
- created 120 *.har files, each about 2.5 MB
- set breakpoints in the three `pipeline.on` callbacks (lines: 73, 78, 96)
- started debugging and stepped through the code by keeping either F5, F10, or F11 continuously pressed.
- did this about 10 times without running into the issue.

Did you do anything differently when you saw the problem?

**Comment by jgsacris** [comment]
On VS Code 1.3.1 I have this problem only when trying to debug mocha tests with the following launch configuration
`{
          "name": "Run mocha",
          "type": "node",
          "request": "launch",
          "program": "${workspaceRoot}/node_modules/mocha/bin/_mocha",
          "stopOnEntry": true,
          "args": [ "--no-timeouts", "--colors",  "test/**/*.test.js"],
          "cwd": "${workspaceRoot}",
          "runtimeExecutable": null,
          "env": { "NODE_ENV": "testing"}
  },`

It stops on entry but on following breakpoints I get the problem mentioned above.
Debugging node running regularly produces no errors.

**Comment by johnfn** [comment]
I wish I could give you guys something more concrete to deal with, but I've found that this issue most often crops up with long running processes. I have a webserver which runs for a long time in vscode. In the first few minutes, I can almost always consistently break on a breakpoint. 

After 10-20 minutes of runtime, more often than not, I run into this issue when trying to debug.

**Comment by rweng** [comment]
I have this issue when having a wildcard in the tests. e.g., I have 4 files:

```
→ l test/integration/
accounts.js     index.js        transactions.js users.js
```

The launch configuration that causes this error is

```js
{
            // Name of configuration; appears in the launch configuration drop down menu.
            "name": "Run mocha",
            // Type of configuration. Possible values: "node", "mono".
            "type": "node",
            "request": "launch",
            // Workspace relative or absolute path to the program.
            "program": "${workspaceRoot}/node_modules/mocha/bin/_mocha",
            // Automatically stop program after launch.
            "stopOnEntry": false,
            // Command line arguments passed to the program.
            "args": [
                "test/integration/*.js"
            ],
            "runtimeArgs": [
                //    "--nolazy"
            ],
            // Workspace relative or absolute path to the working directory of the program being debugged. Default is the current workspace.
            "cwd": "${workspaceRoot}",
            // Workspace relative or absolute path to the runtime executable to be used. Default is the runtime executable on the PATH.
            "runtimeExecutable": null,
            // Environment variables passed to the program.
            "env": {
                "NODE_ENV": "test"
            },
            "sourceMaps": true,
            "outDir": "${workspaceRoot}/dist"
        }
```

But when replacing the wildcard w/

```
"test/integration/index.js",
"test/integration/users.js",
"test/integration/accounts.js",
"test/integration/transactions.js"
```

everything works fine. Hope that helps in nailing it down.

**Comment by juancampa** [comment]
FWIW, I'm getting **almost** the same error exept that it says `vscode_backtrace: Internal Error`, instead of just `Internal Error`.

#### My versions are:
OSX El Capitan 10.11.5
VS Code 1.4.0 6276dcb0ae497766056b4c09ea75be1d76a8b679
Node 6.2.2

**Comment by weinand** [comment]
@juancampa in what situations are you seeing his error? Can you share the project where you experience the problem?

**Comment by johnfn** [comment]
This bug is pretty bad and breaks debugging in a high percentage of cases for me. I can reproduce it consistently; let me know if there's any way I can help you guys track it down.

**Comment by weinand** [comment]
@johnfn it would be great if you could share some code and a project setup where you experience the problem. If we can reproduce the problem, we can investigate and probably fix.
But so far we could not reproduce...

**Comment by johnfn** [comment]
Ahh, that's unfortunate. It's my workplace code, so I can't really share it, and it's rather large (100k loc). 

Is there any other assistance I could give?

My guess is it probably has something to do with deep callstacks, or generators & promises (which we use a lot of).

**Comment by weinand** [comment]
@johnfn use of promises seems to be a common pattern of this issue. What kind of promises are you using?

**Comment by icezohu** [comment]
Hi @weinand yes it happens while using co with generator functions. Now it happens every time I debug generator code blocks.

**Comment by weinand** [comment]
@icezohu what is "co" ?

**Comment by icezohu** [comment]
It's an npm module which could help make the generator code blocks elegant.
https://github.com/tj/co
@weinand

**Comment by johnfn** [comment]
Lots of Bluebird.coroutine, which transforms generators into a sort of async/await pattern. Bluebird is a Promise library for js.

**Comment by pietgk** [comment]
'no call stack available' appeared again this morning in release:
Version 1.5.0-insider (1.5.0-insider)
5fb5131a7d8ab0245af3653fca77f31d7d81d735
2016-09-02T15:32:08.421Z

heavy async context and a breakpoint on gives the 'no call stack available' error, breakpoint off it continues without the 'no call stack available' .

**Comment by Zarel** [comment]
Oh, oh, I have an easily reproducible testcase. At least on my setup (macOS Sierra beta, VSC 1.5.2).

1\. Clone https://github.com/Zarel/flodoc
2\. Checkout 3d6332eea4
3\. Set `launch.json` to

```
{
    "version": "0.2.0",
    "configurations": [
		{
			"name": "Run mocha",
			"type": "node",
			"program": "${workspaceRoot}/node_modules/mocha/bin/_mocha",
			"stopOnEntry": false,
			"args": ["test/spec/*"],
			"cwd": "${workspaceRoot}",
			"runtimeExecutable": null,
			"env": { "NODE_ENV": "development"}
		}
    ]
}
```

4\. Set a breakpoint on line 73 of `src/converter.js`
5\. Debug "Run mocha"
6\. Mash "Continue" a few times
7\. 

![image](https://cloud.githubusercontent.com/assets/551184/18542968/6f4e71d6-7ae2-11e6-8cde-586d32061226.png)

**Comment by pisrael** [comment]
+1 with the problem.
OSX 10.11.6
Visual Studo Code - Version 1.5.3 (1.5.3)

In my case it seems related to bluebird Promise. I removed bluebird from the file I was debugging, using the standard ES6 Promise instead, the error stopped and I was able to debug it.

**Comment by johnfn** [comment]
I've run into it without using Bluebird, but it's significantly more rare.

**Comment by dsbert** [comment]
I can also confirm this issue is happening inside a generator function using bluebird's `co`

**Comment by weinand** [comment]
If you can use node >= v6.3 then please try our new experimental debugger by setting the `type` attribute in your launch config to "node2". That should solve the issue.

**Comment by ssynix** [comment]
Adding `"trace": "7683"` to the launch config did not change anything for me and I still got the same error. Using `node2` seems to have fixed it! 👍

**Comment by weinand** [comment]
@ssynix glad to hear that node2 has fixed this for you.

setting 'trace' to the number of this issue does not make sense ;-)
A better value is "all".

**Comment by ssynix** [comment]
@weinand I was following your suggestion from [earlier on this thread](https://github.com/Microsoft/vscode/issues/7683#issuecomment-233152824). VSCode 1.6.1 complains about it: `Property trace is not allowed`.

Either way, thanks for the suggestion!

**Comment by weinand** [comment]
@ssynix oops, I see. Sorry about the bogus trace value from above. I've fixed my comment.

**Comment by CharlieScarver** [comment]
Thank you! @weinand 
This was **the** most annoying issue ever.

**Comment by cmdrsabre** [comment]
I still habe this issue with an attached debugging session. With `"type":"node2"` it won't attach and `"trace": "all"` does nothing. VSCode 1.7.2, macOS Sierra 10.12.1

**Comment by cmdrsabre** [comment]
@weinand "node2" is not attaching (not used trace)
`Cannot connect to runtime process, timeout after 10000 ms - (reason: Cannot connect to the target: Parse Error).`

I'm running a jasmine testsuite  inside a docker. Breakpoints in the specs work, but not in the application code (other thread), but I think it's more related to the massiv usage of promises.

**Comment by weinand** [comment]
@cmdrsabre make sure that node is launched with the new "--inspect" flag. If it is not, node2 will not be able to attach to it successfully and will timeout with the error you are seeing.

**Comment by cmdrsabre** [comment]
@weinand finally this worked. "node2" worked. Thanks a lot

**Comment by rdewolff** [comment]
Same issue here with last vscode version. 

Trying with ```node2``` and get the following Unhandled error in debug adapter message :

```
******** Unhandled error in debug adapter - Unhandled promise rejection: TypeError: Column must be greater than or equal to 0, got -1
    at SourceMapConsumer_findMapping [as _findMapping] (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/ms-vscode.node-debug2/node_modules/source-map/lib/source-map-consumer.js:543:13)
    at SourceMapConsumer_generatedPositionFor [as generatedPositionFor] (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/ms-vscode.node-debug2/node_modules/source-map/lib/source-map-consumer.js:748:22)
    at SourceMap.generatedPositionFor (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/ms-vscode.node-debug2/node_modules/vscode-chrome-debug-core/out/src/sourceMaps/sourceMap.js:159:34)
    at SourceMaps.mapToGenerated (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/ms-vscode.node-debug2/node_modules/vscode-chrome-debug-core/out/src/sourceMaps/sourceMaps.js:28:18)
    at args.breakpoints.forEach.bp (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/ms-vscode.node-debug2/node_modules/vscode-chrome-debug-core/out/src/transformers/baseSourceMapTransformer.js:63:53)
    at Array.forEach (native)
    at EagerSourceMapTransformer.setBreakpoints (/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/ms-vscode.node-debug2/node_modules/vscode-chrome-debug-core/out/src/transformers/baseSourceMapTransformer.js:61:34)
    at validateBreakpointsPat[...]
```

What's this?

**Comment by roblourens** [comment]
@rdewolff Can you open a new issue for that? That's a new issue and I'd like to look into it when I have time.

**Comment by roblourens** [comment]
Also it would be very helpful if you run with `"diagnosticLogging": true`, then include the full log file from the path printed at the top of the console.

**Comment by smulikHakipod** [comment]
Any updates on this?
I am using runInContext from the [VM module](https://nodejs.org/api/vm.html).
node2 plugin uses node-inspector, which cant debug code running under vm module at all, so I must use the regular node debugger in vscode. In my workspace its crashing all the time (runInContext and BlueBird project). What can I do?
Thanks

**Comment by geogeim** [comment]
Just updated vstudio and ran into this problem

Version 1.9.0
Commit 27240e71ef390bf2d66307e677c2a333cebf75af
Date 2017-02-02T08:31:00.827Z
Shell 1.4.6
Renderer 53.0.2785.143
Node 6.5.0

**Comment by weinand** [comment]
If you can use node >= v6.3 then please try our new experimental debugger by setting the type attribute in your launch config to "node2". That should solve the issue.

**Comment by ilovett** [comment]
was running into this with node 6.9.5 and  vscode Version 1.9.1 (1.9.1)

changed launch config to node2 works

**Comment by bgradin** [comment]
I am also experiencing this error.

VSCode 1.10.1
Mac OSX 10.11.6
node v7.6.0
electron v1.6.1

I am not able to use "node2"; the debugger will not attach, even though electron is being called with `--inspect`. [This issue](https://github.com/Microsoft/vscode/issues/19681) suggests that "node2" simply will not work with electron.

I've noticed that every time I restart my computer, the debugger will hit the breakpoint normally (ie returning a valid callstack and variable information) once and only once. After that, it reverts to the error state and nothing I can do seems to fix it.

It seems like some cached file is in an invalid state. Does vscode have some cache I can clear?

**Comment by edvinv** [comment]
Same with VSCode 1.10.1. 

In my situation it looks like this:
- I set 3 breakpoints and start new debug session. Debugger correctly stops on all three breakpoints. 
- I repeat previous action in the same debug session. Debugger stops on first and second breakpoint. But for third breakpoint it stops on unknown location and I get error "No call stack available (Internal error: TypeError: Cannot read property '0' of undefined)." 
- if I restart debug session it works again for the first time, but not for second, third etc..

The situation is the same for legacy and inspector protocol.

For me breakpoints are practically unusable, because after getting error for the first time, they become totally unreliable.

**Comment by chiangf** [comment]
I'm hitting the same thing on 1.10.2, can we re-open this issue?

**Comment by roblourens** [comment]
Sounds like the new 'inspector protocol' debug adapter fixed this for everyone. I recommend using `"protocol": "inspector"`, if you are using Node v6.9+

**Comment by weinand** [comment]
@chiangf alternatively you could try `"protocol": "legacy"` with the Insiders build (1.11.0). We've fixed an issue in March that might have solved this issue as well.

**Comment by brtthomp** [comment]
I'm having the same issue as @bgradin with electron.  Unfortunately you cannot use node2 with electron so I have to downgrade electron to 1.5.x to get it to work.  Not sure if this is related to electron or vscode.

**Comment by weinand** [comment]
@brtthomp have you tried to use the Insiders release?

**Comment by brtthomp** [comment]
@weinand That seems to have fixed the issue even with electron 1.6.x+.  I didn't read the comments well enough to realize the Insiders build had a potential fix in it.  Thanks!

**Comment by weinand** [comment]
@brtthomp thanks for trying again.
The March stable release will appear soon...

**Comment by jlira2** [comment]
still happening as of now 
VSCode: Version 1.11.2 (1.11.2)
Node: Version v7.7.2

**Comment by weinand** [comment]
@jlira2 please add `"protocol":"inspector"` to your launch config. This should make the problem disappear.

**Comment by michaelkuk** [comment]
Just ran into the same problem, however in my case it was some issue within node module (lodash@4.17) that caused the issue. Downgrading lodash has solved the issue for me.
